crates/release_plz_core/src/command/release.rs

```diff
@@ -874,13 +882,164 @@ struct ReleaseInfo<'a> {
     prs: &'a [Pr],
 }
 
+/// Creates git tag and GitHub release if they don't already exist.
+/// This function is idempotent - it checks if each resource exists before creating.
+/// Returns `true` if any git operation was performed, `false` if everything already existed.
+async fn create_git_tag_and_release(
+    input: &ReleaseRequest,
+    repo: &Repo,
+    git_client: &GitClient,
+    release_info: &ReleaseInfo<'_>,
+) -> anyhow::Result<bool> {
+    if input.dry_run {
+        return Ok(false);
+    }
+
+    let should_create_git_tag = input.is_git_tag_enabled(&release_info.package.name);
+    let should_create_git_release = input.is_git_release_enabled(&release_info.package.name);
+
+    let mut created_something = false;
+
+    // Create git tag if needed and it doesn't exist
+    if should_create_git_tag {
+        // Check if tag exists locally
+        let tag_exists_locally = repo.tag_exists(release_info.git_tag)?;
+
+        if tag_exists_locally {
+            info!(
+                "skipping creation of git tag {}: already exists locally",
+                release_info.git_tag
+            );
+        } else {
```

Comment:
This (combined with other stuff you're doing below) is creating ungodly layers of nesting; can this be extracted to another function? Then you can use the guard pattern.

Reply:
Still applicable, I would personally split this up into multiple functions (`create_git_tag` and `create_git_release` for example), and use the guard pattern within those functions. But that's a matter of style.

Reply:
I hope to find some time in the next couple of days to give this a proper scrub